File: backend/seed.py
import asyncio
import csv
import os
import random
import hashlib
from prisma import Prisma

# Enums
ROLE_USER = "USER"
ROLE_ADMIN = "ADMIN"
STATUS_ACTIVE = "ACTIVE"

# District coordinates mapping fallback
district_coords = {
    'กรุงเทพมหานคร_พระนคร': {'lat': 13.7563, 'lng': 100.5018},
    'ชลบุรี_บางละมุง': {'lat': 12.9236, 'lng': 100.8824},
    'ชลบุรี_ศรีราชา': {'lat': 13.1111, 'lng': 100.9999},
    'นครนายก_เมืองนครนายก': {'lat': 14.2069, 'lng': 101.1965},
    'อุดรธานี_กุมภวาปี': {'lat': 17.1147, 'lng': 103.0181},
    'อุดรธานี_เมืองอุดรธานี': {'lat': 17.4037, 'lng': 102.7895},
    'ขอนแก่น_เมืองขอนแก่น': {'lat': 16.4322, 'lng': 102.8236},
    'สุโขทัย_เมืองสุโขทัย': {'lat': 17.0071, 'lng': 99.8262},
    'อุดรธานี_หนองวัวซอ': {'lat': 17.1654, 'lng': 102.5886},
    'หนองคาย_เมืองหนองคาย': {'lat': 17.8785, 'lng': 102.7423},
    'บึงกาฬ_เมืองบึงกาฬ': {'lat': 18.3614, 'lng': 103.6534},
    'สมุทรปราการ_บางพลี': {'lat': 13.6050, 'lng': 100.7067},
    'เชียงใหม่_เมืองเชียงใหม่': {'lat': 18.7883, 'lng': 98.9853},
}

residential_images = [
    "/images/images (1).jpg",
    "/images/images (2).jpg",
    "/images/images (3).jpg",
    "/images/images (4).jpg",
    "/images/images (5).jpg",
    "/images/images (6).jpg"
]
commercial_images = [
    "/images/images (7).jpg",
    "/images/images (8).jpg",
    "/images/images (9).jpg",
    "/images/images (10).jpg",
    "/images/images (11).jpg",
    "/images/images (12).jpg"
]
industrial_images = [
    "/images/images (13).jpg",
    "/images/images (14).jpg",
    "/images/images (15).jpg",
    "/images/images (16).jpg",
    "/images/images (17).jpg",
    "/images/images (18).jpg"
]
vacant_images = [
    "/images/images (19).jpg",
    "/images/images (20).jpg",
    "/images/images (21).jpg",
    "/images/images (22).jpg",
    "/images/images (23).jpg",
    "/images/images.jpg",
    "/images/dszfgdrhtrj.jpg",
    "/images/esdgdxfh.jpg"
]

def generate_thai_id(name: str) -> str:
    h = hashlib.md5(name.encode('utf-8')).hexdigest()
    val = int(h, 16) % 9000000000000 + 1000000000000
    return f"3{str(val)[1:]}"

def get_random_offset(radius=0.02):
    return (random.random() - 0.5) * radius * 2

def parse_float(val, default=0.0) -> float:
    if val is None:
        return default
    s = str(val).replace('"', '').replace(',', '').strip()
    if not s:
        return default
    try:
        return float(s)
    except ValueError:
        return default

def parse_int(val, default=None):
    if val is None:
        return default
    s = str(val).replace('"', '').replace(',', '').strip()
    if not s:
        return default
    try:
        return int(float(s))
    except ValueError:
        return default

def infer_building_details(lessee_name: str, land_area_sqw: float):
    name = lessee_name.lower()
    if any(x in name for x in ['โรงงาน', 'แปรรูป', 'อุตสาหกรรม']):
        return {
            "building_type": "โรงงาน/คลังสินค้า",
            "usable_area_sqm": float(round(land_area_sqw * 4 * 0.6)),
            "zoning": "พื้นที่สีม่วง (อุตสาหกรรม)"
        }
    elif any(x in name for x in ['อาคาร', 'คูหา', 'ตึก', 'หจก', 'บจก', 'สมาคม', 'สหกรณ์', 'บริษัท', 'ตลาด', 'complex']):
        return {
            "building_type": "อาคารพาณิชย์",
            "usable_area_sqm": float(round(land_area_sqw * 4 * 1.2)),
            "zoning": "พื้นที่สีแดง (พาณิชยกรรม)"
        }
    elif any(x in name for x in ['นาย', 'นาง', 'เด็ก', 'นางสาว']):
        return {
            "building_type": "บ้านพักอาศัย",
            "usable_area_sqm": float(round(land_area_sqw * 4 * 0.4)),
            "zoning": "พื้นที่สีเหลือง (ที่อยู่อาศัยหนาแน่นน้อย)"
        }
    else:
        return {
            "building_type": "ที่ดินเปล่า",
            "usable_area_sqm": 0.0,
            "zoning": "พื้นที่สีเขียว (เกษตรกรรม)"
        }

def get_asking_price(building_type: str, land_area: float) -> float:
    if building_type == "โรงงาน/คลังสินค้า":
        multiplier = 15000
    elif building_type == "อาคารพาณิชย์":
        multiplier = 20000
    elif building_type == "บ้านพักอาศัย":
        multiplier = 8000
    else:
        multiplier = 2000
    base = max(land_area, 10.0) * multiplier * random.uniform(0.95, 1.15)
    return float(round(base / 10000) * 10000)

def generate_description(province: str, district: str, sub_district: str, land_area: float, building_type: str) -> str:
    if building_type == "อาคารพาณิชย์":
        return f"สิทธิ์การเช่าอาคารพาณิชย์ ทำเลดีในย่านชุมชน {sub_district} อำเภอ{district} จังหวัด{province} เนื้อที่ {land_area} ตร.ว. เหมาะสำหรับเปิดร้านค้า สำนักงาน หรือโชว์รูม เดินทางสะดวกติดถนนใหญ่สภาพแวดล้อมดีเยี่ยม"
    elif building_type == "บ้านพักอาศัย":
        return f"สิทธิ์การเช่าที่ดินพร้อมสิ่งปลูกสร้างบ้านพักอาศัย บรรยากาศเงียบสงบในพื้นที่ {sub_district} อำเภอ{district} จังหวัด{province} เนื้อที่ {land_area} ตร.ว. สภาพแวดล้อมดี ปลอดภัย เหมาะสำหรับครอบครัวหรือทำบ้านพักตากอากาศ"
    elif building_type == "โรงงาน/คลังสินค้า":
        return f"สิทธิ์การเช่าคลังสินค้าโรงงานอุตสาหกรรม โครงสร้างแข็งแรงในพื้นที่ {sub_district} อำเภอ{district} จังหวัด{province} เนื้อที่ {land_area} ตร.ว. รองรับการผลิต จัดเก็บสินค้า และการกระจายสินค้า มีทางเข้าออกสำหรับรถบรรทุกสะดวกสบาย"
    else:
        return f"แปลงที่ดินเปล่าราชพัสดุในทำเลศักยภาพ {sub_district} อำเภอ{district} จังหวัด{province} เนื้อที่ {land_area} ตร.ว. ดินดี น้ำอุดมสมบูรณ์ เหมาะสำหรับทำเกษตรกรรม หรือพัฒนาโครงการเชิงพาณิชย์และที่อยู่อาศัย"

async def main():
    db = Prisma()
    await db.connect()
    
    print("Cleaning database...")
    await db.listing.delete_many()
    await db.leasecontract.delete_many()
    await db.user.delete_many()
    
    print("Creating core default users...")
    default_users = [
        {"thaid_id": "1123456789012", "first_name": "สมชาย", "last_name": "ใจดี", "phone_number": "0812345678", "role": ROLE_USER},
        {"thaid_id": "2123456789012", "first_name": "สมหญิง", "last_name": "รักดี", "phone_number": "0898765432", "role": ROLE_USER},
        {"thaid_id": "3123456789012", "first_name": "ประยุทธ์", "last_name": "มั่งมี", "phone_number": "0811112222", "role": ROLE_USER},
        {"thaid_id": "9123456789012", "first_name": "แอดมิน", "last_name": "ธนารักษ์", "phone_number": "021234567", "role": ROLE_ADMIN}
    ]
    
    user_map = {}
    for user_data in default_users:
        user = await db.user.create(data=user_data)
        user_map[user_data["thaid_id"]] = user

    # Candidate CSV file paths
    csv_paths = [
        "data/data.csv",
        "../data/data.csv",
        "backend/data/data.csv",
        "c:/TRD_lex/data/data.csv",
        "./data.csv",
        "datatest.csv",
        "../datatest.csv"
    ]
    csv_file = None
    for path in csv_paths:
        if os.path.exists(path):
            csv_file = path
            break
            
    if not csv_file:
        print("ERROR: data.csv not found!")
        await db.disconnect()
        return
        
    print(f"Reading data from: {csv_file}")
    
    listings_count = 0
    contracts_count = 0
    
    with open(csv_file, mode='r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        
        for row in reader:
            primary_key = row.get("PrimaryKey", "").strip()
            reg_id = row.get("REG_ID", "").strip()
            parcel_num = row.get("Land_No", "").strip()
            rent_cat = row.get("Rent", "").strip()
            rent_type = row.get("Rent_Type", "").strip()
            contract_num = row.get("Rent_Num", "").strip() or primary_key
            vpn_code = row.get("VPN_Code", "").strip()
            lessee_full = row.get("Rent_Name", "").strip() or row.get("ชื่อผู้เช่า/หน่วยงานครอบครอง", "").strip()
            
            if not contract_num and not primary_key:
                continue
                
            sub_district = row.get("Tambon", "").strip() or row.get("ตำบล", "").strip()
            district = row.get("Amphoe", "").strip() or row.get("อำเภอ", "").strip()
            province = row.get("Provice", "").strip() or row.get("Province", "").strip() or row.get("จังหวัด", "").strip()
            
            area_rai = parse_float(row.get("Area_Rai", "0"))
            area_ngan = parse_float(row.get("Area_Ngan", "0"))
            area_wa = parse_float(row.get("Area_Wa", "0"))
            land_area = parse_float(row.get("เนื้อที่ (ตร.ว.)", "0"))
            if land_area == 0:
                land_area = area_rai * 400.0 + area_ngan * 100.0 + area_wa
                
            lat = parse_float(row.get("Latitude", "0"))
            lng = parse_float(row.get("Longitude", "0"))
            
            if lat == 0 or lng == 0:
                loc_key = f"{province}_{district}"
                coords = district_coords.get(loc_key, {'lat': 17.4037, 'lng': 102.7895})
                lat = coords['lat'] + get_random_offset()
                lng = coords['lng'] + get_random_offset()
                
            land_plan = row.get("Land_Plan", "").strip()
            build_type = row.get("Build_Type", "").strip()
            building = row.get("Building", "").strip()
            build_area = parse_float(row.get("Build_Area", "0"))
            build_year = parse_int(row.get("Build_Year"))
            on_street = row.get("ON_Street", "").strip()
            street_t = row.get("Street_T", "").strip()
            street_w = parse_float(row.get("Street_W"))
            street_a = parse_float(row.get("Street_A"))
            land_w = parse_float(row.get("Land_Width"))
            land_ap = parse_float(row.get("Land_AP"))
            build_ap = parse_float(row.get("Build_AP"))
            status = row.get("Status", "ACTIVE").strip().upper()
            
            names = lessee_full.split(" ", 1) if lessee_full else ["ผู้เช่า", "ราชพัสดุ"]
            first_name = names[0]
            last_name = names[1] if len(names) > 1 else ""
            
            thaid_id = generate_thai_id(lessee_full) if lessee_full else f"3{random.randint(100000000000, 999999999999)}"
            
            # Map first 3 active contracts to core test accounts for easy testing
            if contracts_count == 0:
                thaid_id = "1123456789012"
            elif contracts_count == 1:
                thaid_id = "2123456789012"
            elif contracts_count == 2:
                thaid_id = "3123456789012"
                
            if thaid_id in user_map:
                seller = user_map[thaid_id]
            else:
                seller = await db.user.create(
                    data={
                        "thaid_id": thaid_id,
                        "first_name": first_name,
                        "last_name": last_name,
                        "phone_number": f"08{random.randint(10000000, 99999999)}",
                        "role": ROLE_USER
                    }
                )
                user_map[thaid_id] = seller
                
            bld_info = infer_building_details(lessee_full, land_area)
            building_type = build_type if build_type else bld_info["building_type"]
            usable_area_sqm = build_area if build_area > 0 else bld_info["usable_area_sqm"]
            zoning = land_plan if land_plan else bld_info["zoning"]
            
            contracts_count += 1
            
            # Create contract with all new columns
            contract = await db.leasecontract.create(
                data={
                    "contract_number": contract_num,
                    "lessee_thaid_id": thaid_id,
                    "parcel_number": parcel_num,
                    "location_lat": lat,
                    "location_lng": lng,
                    "province": province,
                    "district": district,
                    "sub_district": sub_district,
                    "land_area_sqw": land_area,
                    "is_active": (status != "HIDDEN"),
                    "building_type": building_type,
                    "usable_area_sqm": usable_area_sqm,
                    "zoning": zoning,
                    "annual_rent": 12000.0 if land_ap == 0 else float(land_ap),
                    
                    # Additional CSV fields
                    "primary_key": primary_key,
                    "reg_id": reg_id,
                    "rent_category": rent_cat,
                    "rent_type": rent_type,
                    "vpn_code": vpn_code,
                    "lessee_name": lessee_full,
                    "area_rai": area_rai,
                    "area_ngan": area_ngan,
                    "area_wa": area_wa,
                    "land_plan": land_plan,
                    "building_details": building,
                    "build_year": build_year,
                    "on_street": on_street,
                    "street_type": street_t,
                    "street_width": street_w,
                    "street_access": street_a,
                    "land_width": land_w,
                    "land_ap": land_ap,
                    "build_ap": build_ap
                }
            )
            
            # Create active listing for first 50 contracts or all active ones
            if status == STATUS_ACTIVE and listings_count < 60:
                asking_price = get_asking_price(building_type, land_area)
                estimated_fee = asking_price * 0.03
                description = generate_description(province, district, sub_district, land_area, building_type)
                
                if building_type == "อาคารพาณิชย์" or "ตึก" in building:
                    images = commercial_images
                elif "โรงงาน" in building_type or "คลัง" in building_type:
                    images = industrial_images
                elif "บ้าน" in building_type or "อาศัย" in building_type:
                    images = residential_images
                else:
                    images = vacant_images
                    
                img_url = random.choice(images)
                
                listings_count += 1
                await db.listing.create(
                    data={
                        "id": f"list-{listings_count}",
                        "sellerId": seller.id,
                        "contractId": contract.id,
                        "asking_price": asking_price,
                        "estimated_fee": estimated_fee,
                        "description": description,
                        "image_urls": [img_url],
                        "status": STATUS_ACTIVE
                    }
                )
                
        print(f"Successfully seeded {contracts_count} contracts and {listings_count} active listings from CSV!")
        await db.disconnect()

if __name__ == "__main__":
    asyncio.run(main())

File: backend/prisma/seed.ts
import { PrismaClient, Role, ListingStatus } from '@prisma/client';
import * as fs from 'fs';
import * as path from 'path';

// Load .env file from root directory
const envPath = path.join(__dirname, '../../.env');
if (fs.existsSync(envPath)) {
  const envContent = fs.readFileSync(envPath, 'utf-8');
  for (const line of envContent.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const firstEq = trimmed.indexOf('=');
    if (firstEq === -1) continue;
    const key = trimmed.substring(0, firstEq).trim();
    let val = trimmed.substring(firstEq + 1).trim();
    if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }
    process.env[key] = val;
  }
}

const prisma = new PrismaClient();

// Deterministic mock Thai ID generator
function generateThaiIdFromName(name: string): string {
  let hash = 0;
  for (let i = 0; i < name.length; i++) {
    hash = name.charCodeAt(i) + ((hash << 5) - hash);
  }
  const idNum = (Math.abs(hash) % 9000000000000) + 1000000000000;
  // Standard citizen IDs in Thailand often start with 3
  return `3${idNum.toString().slice(1)}`;
}

// Returns a random offset within a given degree radius (default 0.03 degrees, approx 3-5 km)
function getRandomOffset(radius = 0.03): number {
  return (Math.random() - 0.5) * radius * 2;
}

// Infer building details based on lessee name
function inferBuildingDetails(lesseeName: string, landAreaSqw: number): { buildingType: string; usableAreaSqm: number; zoning: string } {
  const name = lesseeName.toLowerCase();
  
  if (name.includes('โรงงาน') || name.includes('แปรรูป') || name.includes('อุตสาหกรรม')) {
    return {
      buildingType: 'โรงงาน/คลังสินค้า',
      usableAreaSqm: Math.round(landAreaSqw * 4 * 0.6), // Assume 60% building coverage
      zoning: 'พื้นที่สีม่วง (อุตสาหกรรม)'
    };
  } else if (
    name.includes('อาคาร') || 
    name.includes('คูหา') || 
    name.includes('ตึก') || 
    name.includes('หจก') || 
    name.includes('บจก') || 
    name.includes('สมาคม') || 
    name.includes('สหกรณ์') || 
    name.includes('บริษัท') || 
    name.includes('ตลาด') || 
    name.includes('complex')
  ) {
    return {
      buildingType: 'อาคารพาณิชย์',
      usableAreaSqm: Math.round(landAreaSqw * 4 * 1.2), // Assume 2-3 stories
      zoning: 'พื้นที่สีแดง (พาณิชยกรรม)'
    };
  } else if (name.includes('นาย') || name.includes('นาง') || name.includes('เด็ก')) {
    return {
      buildingType: 'บ้านพักอาศัย',
      usableAreaSqm: Math.round(landAreaSqw * 4 * 0.4), // Assume 40% building coverage
      zoning: 'พื้นที่สีเหลือง (ที่อยู่อาศัยหนาแน่นน้อย)'
    };
  } else {
    // Default to vacant land if it's reserved, public area, or empty
    return {
      buildingType: 'ที่ดินเปล่า',
      usableAreaSqm: 0,
      zoning: 'พื้นที่สีเขียว (เกษตรกรรม)'
    };
  }
}

// District center coordinates mapping
const districtCoords: Record<string, { lat: number; lng: number }> = {
  'กรุงเทพมหานคร_พระนคร': { lat: 13.7563, lng: 100.5018 },
  'ชลบุรี_บางละมุง': { lat: 12.9236, lng: 100.8824 },
  'ชลบุรี_ศรีราชา': { lat: 13.1111, lng: 100.9999 },
  'นครนายก_เมืองนครนายก': { lat: 14.2069, lng: 101.1965 },
  'อุดรธานี_กุมภวาปี': { lat: 17.1147, lng: 103.0181 },
  'ขอนแก่น_เมืองขอนแก่น': { lat: 16.4322, lng: 102.8236 },
  'สุโขทัย_เมืองสุโขทัย': { lat: 17.0071, lng: 99.8262 },
  'อุดรธานี_หนองวัวซอ': { lat: 17.1654, lng: 102.5886 },
  'หนองคาย_เมืองหนองคาย': { lat: 17.8785, lng: 102.7423 },
  'บึงกาฬ_เมืองบึงกาฬ': { lat: 18.3614, lng: 103.6534 },
  'สมุทรปราการ_บางพลี': { lat: 13.6050, lng: 100.7067 },
  'เชียงใหม่_เมืองเชียงใหม่': { lat: 18.7883, lng: 98.9853 },
};

// Robust CSV parser function
function parseCSV(content: string): string[][] {
  const result: string[][] = [];
  let row: string[] = [];
  let cell = '';
  let inQuotes = false;

  for (let i = 0; i < content.length; i++) {
    const char = content[i];
    const nextChar = content[i + 1];

    if (inQuotes) {
      if (char === '"') {
        if (nextChar === '"') {
          cell += '"';
          i++; // Skip next quote
        } else {
          inQuotes = false;
        }
      } else {
        cell += char;
      }
    } else {
      if (char === '"') {
        inQuotes = true;
      } else if (char === ',') {
        row.push(cell);
        cell = '';
      } else if (char === '\n' || char === '\r') {
        if (char === '\r' && nextChar === '\n') {
          i++;
        }
        row.push(cell);
        result.push(row);
        row = [];
        cell = '';
      } else {
        cell += char;
      }
    }
  }

  if (cell !== '' || row.length > 0) {
    row.push(cell);
    result.push(row);
  }

  return result;
}

async function main() {
  console.log('Cleaning database...');
  // Clean up data in reverse order of relationships
  await prisma.listing.deleteMany();
  await prisma.leaseContract.deleteMany();
  await prisma.user.deleteMany();

  console.log('Seeding default users...');
  const userSomchai = await prisma.user.create({
    data: {
      thaid_id: '1123456789012',
      first_name: 'สมชาย',
      last_name: 'ใจดี',
      phone_number: '0812345678',
      role: Role.USER,
    },
  });

  const userSomying = await prisma.user.create({
    data: {
      thaid_id: '2123456789012',
      first_name: 'สมหญิง',
      last_name: 'รักดี',
      phone_number: '0898765432',
      role: Role.USER,
    },
  });

  const userAdmin = await prisma.user.create({
    data: {
      thaid_id: '9123456789012',
      first_name: 'แอดมิน',
      last_name: 'ธนารักษ์',
      phone_number: '021234567',
      role: Role.ADMIN,
    },
  });

  console.log('Seeding default lease contracts...');
  const contract1 = await prisma.leaseContract.create({
    data: {
      contract_number: 'TRD-66-001',
      lessee_thaid_id: '1123456789012', // Somchai
      parcel_number: 'กท.1234',
      location_lat: 13.7563,
      location_lng: 100.5018,
      province: 'กรุงเทพมหานคร',
      district: 'พระนคร',
      sub_district: 'วัดราชบพิธ',
      land_area_sqw: 50.0,
      is_active: true,
      building_type: 'อาคารพาณิชย์',
      usable_area_sqm: 120.0,
      zoning: 'พื้นที่สีแดง (พาณิชยกรรม)',
    },
  });

  const contract2 = await prisma.leaseContract.create({
    data: {
      contract_number: 'TRD-66-002',
      lessee_thaid_id: '2123456789012', // Somying
      parcel_number: 'ชบ.5678',
      location_lat: 12.9236,
      location_lng: 100.8824,
      province: 'ชลบุรี',
      district: 'บางละมุง',
      sub_district: 'หนองปรือ',
      land_area_sqw: 120.0,
      is_active: true,
      building_type: 'บ้านพักอาศัย',
      usable_area_sqm: 180.0,
      zoning: 'พื้นที่สีเหลือง (ที่อยู่อาศัยหนาแน่นน้อย)',
    },
  });

  const contract3 = await prisma.leaseContract.create({
    data: {
      contract_number: 'TRD-66-003',
      lessee_thaid_id: '1123456789012', // Somchai (Inactive)
      parcel_number: 'ชบ.9999',
      location_lat: 13.1111,
      location_lng: 100.9999,
      province: 'ชลบุรี',
      district: 'ศรีราชา',
      sub_district: 'ศรีราชา',
      land_area_sqw: 80.0,
      is_active: false,
      building_type: 'ที่ดินเปล่า',
      usable_area_sqm: 0.0,
      zoning: 'พื้นที่สีเขียว (เกษตรกรรม)',
    },
  });

  console.log('Seeding default listings...');
  await prisma.listing.create({
    data: {
      sellerId: userSomchai.id,
      contractId: contract1.id,
      asking_price: 500000.0,
      estimated_fee: 15000.0,
      description: 'สิทธิ์การเช่าที่ดินราชพัสดุทำเลทอง ใจกลางกรุงเทพฯ ใกล้ MRT พระนคร เหมาะกับการเปิดร้านอาหารหรือร้านกาแฟ',
      image_urls: ['https://images.unsplash.com/photo-1500382017468-9049fed747ef?w=800'],
      status: ListingStatus.ACTIVE,
    },
  });

  await prisma.listing.create({
    data: {
      sellerId: userSomying.id,
      contractId: contract2.id,
      asking_price: 1200000.0,
      estimated_fee: 36000.0,
      description: 'ที่ดินแปลงสวยในอำเภอบางละมุง พัทยา ชลบุรี เหมาะสำหรับทำที่พักอาศัยหรือพัฒนาธุรกิจส่วนตัว เดินทางสะดวกมีสาธารณูปโภคครบครัน',
      image_urls: ['https://images.unsplash.com/photo-1464822759023-fed622ff2c3b?w=800'],
      status: ListingStatus.ACTIVE,
    },
  });

  // Reading the CSV file
  const csvPath = 'C:\\TRD_lex\\datatest.csv';
  console.log(`Reading CSV file from ${csvPath}...`);

  if (!fs.existsSync(csvPath)) {
    throw new Error(`CSV file not found at ${csvPath}`);
  }

  const fileContent = fs.readFileSync(csvPath, 'utf-8');
  const parsedRows = parseCSV(fileContent);

  // Headers: ลำดับ,เลขที่ราชพัสดุ,เลขที่สัญญา,ชื่อผู้เช่า/หน่วยงานครอบครอง,จังหวัด,อำเภอ,ตำบล,เนื้อที่ (ตร.ว.),สถานะ
  const rows = parsedRows.slice(1);

  console.log(`Found ${rows.length} rows to seed.`);

  let successCount = 0;
  for (const row of rows) {
    if (row.length < 9 || !row[1] || !row[2]) {
      continue; // Skip invalid rows
    }

    const parcelNumber = row[1].trim();
    const contractNumber = row[2].trim();
    const lesseeName = row[3].trim();
    const province = row[4].trim();
    const district = row[5].trim();
    const subDistrict = row[6].trim();
    const landAreaRaw = row[7].trim().replace(/,/g, '');
    const landAreaSqw = parseFloat(landAreaRaw) || 0;
    const status = row[8].trim().toUpperCase();

    // Map status to isActive
    // If status is ACTIVE or SOLD, is_active is true. If HIDDEN, is_active is false.
    const isActive = status !== 'HIDDEN';

    // Generate deterministic mock Thai ID and name parts for the lessee
    const thaidId = generateThaiIdFromName(lesseeName);
    const nameParts = lesseeName.split(/\s+/);
    const firstName = nameParts[0] || lesseeName;
    const lastName = nameParts.slice(1).join(' ') || '-';

    // Geolocation mapping with random offsets within the district
    const districtKey = `${province}_${district}`;
    const baseCoords = districtCoords[districtKey] || districtCoords[`${province}_เมือง${province}`] || { lat: 13.7563, lng: 100.5018 }; // Default to Bangkok
    const locationLat = baseCoords.lat + getRandomOffset(0.03);
    const locationLng = baseCoords.lng + getRandomOffset(0.03);

    // Infer building details based on lessee name
    const { buildingType, usableAreaSqm, zoning } = inferBuildingDetails(lesseeName, landAreaSqw);

    // First ensure the User exists
    await prisma.user.upsert({
      where: { thaid_id: thaidId },
      update: {},
      create: {
        thaid_id: thaidId,
        first_name: firstName,
        last_name: lastName,
        role: Role.USER,
      },
    });

    // Upsert the LeaseContract
    await prisma.leaseContract.upsert({
      where: { contract_number: contractNumber },
      update: {
        lessee_thaid_id: thaidId,
        parcel_number: parcelNumber,
        location_lat: locationLat,
        location_lng: locationLng,
        province,
        district,
        sub_district: subDistrict,
        land_area_sqw: landAreaSqw,
        is_active: isActive,
        building_type: buildingType,
        usable_area_sqm: usableAreaSqm,
        zoning: zoning,
      },
      create: {
        contract_number: contractNumber,
        lessee_thaid_id: thaidId,
        parcel_number: parcelNumber,
        location_lat: locationLat,
        location_lng: locationLng,
        province,
        district,
        sub_district: subDistrict,
        land_area_sqw: landAreaSqw,
        is_active: isActive,
        building_type: buildingType,
        usable_area_sqm: usableAreaSqm,
        zoning: zoning,
      },
    });

    successCount++;
  }

  console.log(`Seeded ${successCount} lease contracts successfully from CSV.`);
  console.log('Database seeding completed successfully!');
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
